### PR TITLE
avoid allocationg ApplicationProtocol in common case

### PR DESCRIPTION
**Pull request by wfurt**
contributes to #68951 
For now, I updated Linux & Windows. Other platforms may be possible.

### CODE REVIEW

File: src/libraries/System.Net.Security/src/System/Net/Security/SslConnectionInfo.cs [hunk 1]
```diff
@@ -5,6 +5,12 @@ namespace System.Net.Security
 {
     internal partial struct SslConnectionInfo
     {
+#pragma warning disable CA1823
+        private static readonly byte[] s_http1 = SslApplicationProtocol.Http11.Protocol.ToArray();
+        private static readonly byte[] s_http2 = SslApplicationProtocol.Http2.Protocol.ToArray();
+        private static readonly byte[] s_http3 = SslApplicationProtocol.Http3.Protocol.ToArray();
```
Comment (davidfowl) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslConnectionInfo.cs:11]: This feels like an opportunity to use the utf8 string literal feature.


cc @stephentoub

File: src/libraries/System.Net.Security/src/System/Net/Security/SslConnectionInfo.Windows.cs [hunk 2]
```diff
@@ -17,7 +17,20 @@ internal partial struct SslConnectionInfo
                 alpnContext.ProtoNegoExt == Interop.ApplicationProtocolNegotiationExt.ALPN &&
                 alpnContext.ProtoNegoStatus == Interop.ApplicationProtocolNegotiationStatus.Success)
             {
-                return alpnContext.Protocol;
+                if (alpnContext.ProtocolIdSize == s_http1.Length && alpnContext.Protocol.SequenceEqual(s_http1))
```
Comment (stephentoub) [inline src/libraries/System.Net.Security/src/System/Net/Security/SslConnectionInfo.Windows.cs:20]: Why does the Windows one check ProtocolIdSize but the Linux one doesn't?

File: src/libraries/Common/src/Interop/Windows/SChannel/Interop.SecPkgContext_ApplicationProtocol.cs [hunk 3]
```diff
@@ -29,13 +29,13 @@ internal unsafe struct SecPkgContext_ApplicationProtocol
...
         public fixed byte ProtocolId[MaxProtocolIdSize];
-        public byte[] Protocol
+        public ReadOnlySpan<byte> Protocol
         {
             get
             {
                 fixed (byte* pid = ProtocolId)
                 {
-                    return new Span<byte>(pid, ProtocolIdSize).ToArray();
+                    return new ReadOnlySpan<byte>(pid, ProtocolIdSize);
```
[the remark this answers is not in the document]
Comment (stephentoub) [reply src/libraries/Common/src/Interop/Windows/SChannel/Interop.SecPkgContext_ApplicationProtocol.cs:38]: Actually, I take that back; this should be ok.

**Review by rzikm — APPROVED** [review]
LGTM, modulo existing comments

File: src/libraries/System.Net.Security/src/System/Net/Security/SslConnectionInfo.Windows.cs (hunk 2 above)
Comment (wfurt) [reply src/libraries/System.Net.Security/src/System/Net/Security/SslConnectionInfo.Windows.cs:20]: The code originally looked different and I did not want to allocate the Span unless we need to. Perhaps that is cheap. Here I have the size available directly so I left the check in place. On Linux, The Interop function returns always Span so the code is somewhat different. I can remove the extra check if we want to.

**Pull request merged**